### PR TITLE
Integrate bare metal support

### PR DESCRIPTION
**Pull request by jiria**
Introduced providerid as per: https://github.com/microsoft/moc/pull/58
Cleans up baremetalhost and baremetalmachine as per: https://github.com/microsoft/moc/pull/54

### CODE REVIEW

**Review by vincepnguyen — DISMISSED** [review]

**Review by romoh — APPROVED** [review]

File: pkg/providerid/providerid.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	if len(split) == 1 {
+		// VMs don't have the host type prefix.
+		return HostTypeVM, split[0], nil
+	}
+
+	hostType := HostType(split[0])
+	machineName := split[1]
+
+	if hostType != HostTypeBareMetal {
+		return "", "", fmt.Errorf("providerID contains unknown host type: %s", string(hostType))
```
Comment (madhanrm) [inline pkg/providerid/providerid.go:65]: General comment: 
try to wrap errors using known errors defined in errors pkg.
You could use NotFound, or define Unknown etc

File: rpc/cloudagent/compute/baremetalhost/moc_cloudagent_baremetalhost.proto [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+message BareMetalHostNetworkInterface {
+	string networkInterfaceName = 1;
+}
+
+message BareMetalHostNetworkConfiguration {
+	repeated BareMetalHostNetworkInterface interfaces = 1;
+}
+
+message BareMetalHostSize {
+	int32 cpuCount = 1;
```
Comment (madhanrm) [inline rpc/cloudagent/compute/baremetalhost/moc_cloudagent_baremetalhost.proto:47]: can we keep this as uint

File: rpc/cloudagent/compute/baremetalhost/moc_cloudagent_baremetalhost.proto [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+message BareMetalHost {
+	string name = 1;
+	string id = 2;
+	BareMetalHostStorageConfiguration storage = 3;
+	BareMetalHostNetworkConfiguration network = 4;
+	BareMetalHostHardwareConfiguration hardware = 5;
+	BareMetalHostPowerState powerState = 6;
+	SecurityConfiguration security = 7;
+	string bareMetalMachineName = 8;
```
Comment (madhanrm) [inline rpc/cloudagent/compute/baremetalhost/moc_cloudagent_baremetalhost.proto:64]: is this the hostName? can you keep this as MachineName. BareMetal is redundant

Comment (madhanrm) [reply rpc/cloudagent/compute/baremetalhost/moc_cloudagent_baremetalhost.proto:64]: or is this reference to the bareMetalMachine - which doesnt seem right - to have this cyclic dependency.

You can maintain these in the internal structures, when machine gets added to host via AddResource

File: rpc/common/moc_common_common.proto [hunk 4]
```diff
@@ -145,6 +145,7 @@ enum ProviderType {
 	CredentialMonitor      = 31;
 	Logging                = 32;
 	Recovery               = 33;
```
Comment (madhanrm) [inline rpc/common/moc_common_common.proto:147]: Revendor here is required.

File: rpc/cloudagent/compute/baremetalhost/moc_cloudagent_baremetalhost.proto (hunk 3 above)
Comment (cwize1) [reply rpc/cloudagent/compute/baremetalhost/moc_cloudagent_baremetalhost.proto:64]: This is a reference to `bareMetalMachine`. It is a read-only field that is populated by MOC for convenience and debugging.

Comment (madhanrm) [reply rpc/cloudagent/compute/baremetalhost/moc_cloudagent_baremetalhost.proto:64]: There are patterns to do this - look at how group manages the dependencies in groupsInternal structure

Lets follow the same pattern.

Comment (jiria) [reply rpc/cloudagent/compute/baremetalhost/moc_cloudagent_baremetalhost.proto:64]: We have the references in both Internal objects: https://github.com/microsoft/wssdcloudagent/blob/04751dc1fc85c66b5668d7148ae6cb5b455824d1/pkg/cloudagent/services/compute/baremetalhost/internal/baremetalhost.go#L16 and https://github.com/microsoft/wssdcloudagent/blob/04751dc1fc85c66b5668d7148ae6cb5b455824d1/pkg/cloudagent/services/compute/baremetalmachine/internal/baremetalmachine.go#L16. 

With that, do you suggest to delete references from both MOC objects (BareMetalHost and BareMetalMachine) and just keep both references in their Internal objects or are you suggesting something else?

Comment (madhanrm) [reply rpc/cloudagent/compute/baremetalhost/moc_cloudagent_baremetalhost.proto:64]: reference is required in proto, only if it needs to be input by the client.
For debugging purpose and other internal notification purpose, internal object implementation is good.

Unless required, do not expose properties in proto.

Comment (jiria) [reply rpc/cloudagent/compute/baremetalhost/moc_cloudagent_baremetalhost.proto:64]: With that, should any discovered information about BMH be only part of the internal object? Say, CPU count.

Comment (madhanrm) [reply rpc/cloudagent/compute/baremetalhost/moc_cloudagent_baremetalhost.proto:64]: https://github.com/microsoft/moc/blob/main/rpc/cloudagent/cloud/node/moc_cloudagent_node.proto#L39
https://github.com/microsoft/moc/blob/main/rpc/cloudagent/storage/container/moc_cloudagent_container.proto#L37

some sample code on how to propagate node info that is discovered

Comment (jiria) [reply rpc/cloudagent/compute/baremetalhost/moc_cloudagent_baremetalhost.proto]: Re info, I think I want to leave this to another time. I think as we add inspection support, these will need to be modified as well and so would rather do it in one go.

**Review by madhanrm — APPROVED** [review]

File: pkg/providerid/providerid.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+	"github.com/microsoft/moc/pkg/errors"
+)
+
+const (
+	ProviderName     string = "moc"
+	ProviderIDPrefix string = ProviderName + "://"
+)
+
+type HostType string
```
Comment (madhanrm) [inline pkg/providerid/providerid.go:17]: these enums can be defined in Proto, if required.

**Review by madhanrm — APPROVED** [review]

File: pkg/providerid/providerid.go (hunk 5 above)
Comment (jiria) [reply pkg/providerid/providerid.go:17]: @cwize1 FYI

**Pull request merged**